More plot/widget smoke tests for Python and R

testlabauto opened the pull request:
Python:
- bqplot
- ipydatagrid
- ipyleaflet
- ipytree

R:
- rplot
- highcharter
- leaflet
- plotly

Most of these plots render in a nested frame.  These frames don't appear to go away when you clear the plots, so the selector gets the "last" frames as they appear to be added in order.

Note that previously missing packages were added to qa-example-content:
https://github.com/posit-dev/qa-example-content/pull/19

Note that I had to clear up some formatting issues to get around commit problems.

### QA Notes

All smoke tests should pass

testlabauto committed:
Add various Python widget plot tests

testlabauto committed:
More plot/widget tests

testlabauto committed:
capture screenshot when shiny install fails

testlabauto committed:
Merge branch 'main' into cmead/more-plots

testlabauto commented on test/automation/src/positron/positronPlots.ts:
```diff
@@ -52,6 +54,10 @@ export class PositronPlots {
 
+	async waitForWebviewPlot(selector: string) {
+		await this.code.driver.getFrame(OUTER_WEBVIEW_FRAME).last().frameLocator(INNER_WEBVIEW_FRAME).last().locator(selector).waitFor({ state: 'visible' });
+	}
```
Getting the last frame for inner and outer because when you clear a plot, the frame seems to stay behind.  Fortunately, new frames seem to be added sequentially.

seeM replied:
I think it's a bug that we're not properly disposing webview plots. Fixing that would also let us add `waitForNoPlots()` to all test cases.

testlabauto replied:
Note that Wasim has put in a fix for this.

testlabauto commented on test/smoke/src/areas/positron/plots/plots.test.ts:
```diff
@@ -110,9 +110,9 @@ import IPython
-    "C",
+	"A",
+	"B",
+	"C",
```
VSCode wanted this formatting cleaned up.  Not sure it picked me.

testlabauto commented on test/smoke/src/areas/positron/plots/plots.test.ts:
```diff
@@ -153,9 +153,9 @@ import IPython
-    "C",
+	"A",
+	"B",
+	"C",
```
VSCode wanted this formatting cleaned up. Not sure it picked me.

testlabauto commented on test/smoke/src/areas/positron/plots/plots.test.ts:
```diff
@@ -283,6 +283,101 @@ plt.show()`;
+				await app.workbench.positronPlots.waitForWebviewPlot('.svg-figure');
+
+				await app.workbench.positronPlots.clearPlots();
+
```
Not waiting for plots to be gone because structure used to identify plots remains.

seeM committed:
fix webview plot disposal; update smoke tests to wait for plots to clear

seeM commented on test/smoke/src/areas/positron/plots/plots.test.ts:
```diff
@@ -283,6 +283,101 @@ plt.show()`;
 				await app.workbench.positronPlots.waitForNoPlots();
 			});
+
+			it('Python - Verifies bqplot Python widget [C720869]', async function () {
```
Since the code for each of these test cases is the same, should we extract and reuse a function?

testlabauto replied:
Good idea.  Done!

seeM reviewed (commented):
LGTM and tests pass locally. I think you've discovered a bug where we're not properly disposing webview plots. I've fixed that locally and also added all of the `waitForNoPlots()` calls which would currently fail. Do you mind if I push that to your branch?

BTW I think `graphviz` may need to be added to the example content repo's `requirements.txt` file. I also needed to do `brew install graphviz`.

What's the best way to install all of the requirements before running tests?

It's worth noting that all of the Python cases here are for libraries built on top of `ipywidgets`, so they would typically all break together. There are also other widget libraries that are handled differently internally e.g. altair, bokeh, and plotly.

We may also want to try testing the interactivity of some of these widgets, but I imagine that might be a bit trickier.

testlabauto committed:
Factor out common test bodies into reuseable function

testlabauto commented:
> LGTM and tests pass locally. I think you've discovered a bug where we're not properly disposing webview plots. I've fixed that locally and also added all of the `waitForNoPlots()` calls which would currently fail. Do you mind if I push that to your branch?
> 
> BTW I think `graphviz` may need to be added to the example content repo's `requirements.txt` file. I also needed to do `brew install graphviz`.
> 
> What's the best way to install all of the requirements before running tests?
> 
> It's worth noting that all of the Python cases here are for libraries built on top of `ipywidgets`, so they would typically all break together. There are also other widget libraries that are handled differently internally e.g. altair, bokeh, and plotly.
> 
> We may also want to try testing the interactivity of some of these widgets, but I imagine that might be a bit trickier.

Awesome.  Glad to have discovered a bug!  Thanks for getting it fixed!

Note that graphviz is installed/setup here: 
- https://github.com/posit-dev/positron/blob/main/.github/workflows/positron-full-test.yml#L110 
- https://github.com/posit-dev/positron/blob/main/.github/workflows/positron-full-test.yml#L118
(and also the same for the PR/merge workflow)

I am a little on the fence about this approach, but the dependencies are largely kept in qa-example-content:
- https://github.com/posit-dev/qa-example-content/blob/main/requirements.txt
- https://github.com/posit-dev/qa-example-content/blob/main/DESCRIPTION
So, using those files to install dependencies would be the way to go, unless the dependency is only a dependency of the smoke tests and not of qa-example-content.  In that case, you find it here:
- https://github.com/posit-dev/positron/blob/main/.github/workflows/positron-full-test.yml#L118

Oh, I will plan to add tests for other widget libraries as I see tickets come through where you and Nick are fixing things up!

seeM reviewed (approved):
LGTM!
